DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/board_test/src/usm_speed.hpp

```diff
@@ -0,0 +1,177 @@
+class USMMemCopy;
+class USMMemRead;
+class USMMemWrite;
+
+using long8 = sycl::vec<long, 8>;
+
+enum USMTest { MEMCOPY, READ, WRITE };
+constexpr size_t kIterations = 1;
+constexpr size_t kNumBytes = 1024 * 1024 * 1024;
+constexpr size_t kNumItems = kNumBytes / sizeof(long8);
+
+// MEMCOPY test: launches a kernel to copy data from one USM pointer to another.
+sycl::event memcopy_kernel(sycl::queue &q, long8 *in, long8 *out) {
+  return q.single_task<USMMemCopy>([=]() [[intel::kernel_args_restrict]] {
+    sycl::host_ptr<long8> in_h(in);
+    sycl::host_ptr<long8> out_h(out);
+    for (size_t i = 0; i < kNumItems; i++) {
+      out_h[i] = in_h[i];
+    }
+  });
+}
+
+// READ test: launches a kernel to read data from a USM pointer, sum it up, and
+// store to an output pointer.
+sycl::event read_kernel(sycl::queue &q, long8 *in, long8 *out) {
+  return q.single_task<USMMemRead>([=]() {
+    sycl::host_ptr<long8> in_h(in);
+    sycl::host_ptr<long8> out_h(out);
+    long8 sum{0};
+    for (size_t i = 0; i < kNumItems; i++) {
+      sum += in_h[i];
+    }
+    // This prevents the reads from being optimized away
+    out_h[0] = sum;
+  });
+}
+
+// WRITE test: launches a kernel to write data to a USM pointer.
+sycl::event write_kernel(sycl::queue &q, long8 *in, long8 *out) {
+  return q.single_task<USMMemWrite>([=]() {
+    sycl::host_ptr<long8> out_h(out);
+    long8 answer{5};
+    for (size_t i = 0; i < kNumItems; i++) {
+      out_h[i] = answer;
+    }
+  });
+}
+
+// Function to check the answer of the memcopy, read or write test.
+bool verify(long8 *in, long8 *out, USMTest test) {
+
+  // The write kernel writes a known value to every index of "out".
+  long8 write_answer{5};
+
+  // The read kernel calculates a sum of all the values at "in" and stores it
+  // at out[0]. First calculate a reference to compare to. The rest of the
+  // indices should retain their intial value, 0. So, also define a zero vector
+  // to compare to.
+  long8 zero_vec{0};
+  long8 read_answer{0};
+  for (int i = 0; i < kNumItems; i++) {
+    read_answer += in[i];
+  }
+
+  for (int i = 0; i < kNumItems; i++) {
+    // for each element of the output pointer, the value we expect it to take.
+    long8 reference;
+
+    switch (test) {
+    case MEMCOPY:
+      // output pointer should match the input pointer.
+      reference = in[i];
+      break;
+    case READ:
+      // first index should contain the expected answer (sum of all values at
+      // the input pointer); at all other it should contain zero.
+      reference = (i == 0) ? read_answer : zero_vec;
+      break;
+    case WRITE:
+      // each index should contain the known answer that was written.
+      reference = write_answer;
+      break;
+    }
+
+    // "compare" will be a vector containing an element-wise "==" of the
+    // vectors we are checking equality for.
+    long8 compare = (out[i] == reference);
```

Comment:
When I was looking at this code sample,  this compare logic always confused me. It seems way to complex for somethign as simple as checking if an element is equal to reference. I think it'll be better if it was simpler!

Reply:
To add on, I think it was done the way it was in the regtest because it was using template functions, and the type of data was templated. Here, the data is hardcoded as a long8, so you can use that for a simpler comparison

Reply:
Agreed. Rather than creating a separate "compare" vector, I've updated the sample to just compare the two vectors directly.